_posts/2015-06-09-Fostering-Creativity-with-Semantically-Rich-APIs.md

```diff
@@ -0,0 +1,115 @@
+---
+title: Semantically Rich APIs
+excerpt: Fostering Creativity with Semantically Rich APIs
+layout: post
+date: 2015-06-09 00:00:00 +0200
+author: stephen
+published: true
+---
+
+In a recent post on our blog, Kyle Fuller [wrote about](http://blog.apiary.io/2015/06/04/Hyperdrive/) the benefits of decoupling semantics from implementation for building better API clients. This post aims to explore how this decoupled approach also benefits the API design process, specifically around freeing the produces and consumers to be more creative with semantically rich APIs.
+
+## The Difficulty with Common Approaches
+
+Many modern approaches in designing and building APIs couple not only clients and servers to implementation details, but we also couple thought, creativity, and shared understanding to that implementation. Coupling servers and clients is a problem that is easily seen, and in theory no one wants a coupled approach. Kyle Fuller wrote in his post:
+
+> Imagine entering an API without knowing about any implementation details, but instead just understanding the semantics of the API. We can get the implementation details of the API at run-time to allow us to change them at any point.
+
+In waiting until runtime to get these implementation details, Hyperdrive allows for building clients that are highly evolvable and don't break when things are changed on the server.
+
+Coupling hinders API evolvability—among other things—but it also hinders API design and creativity for designers and consumers. It is a subtle impedance we overlook because of the distance we've made it in HTTP APIs and because the spectrum of frameworks, libraries, and tools promote this coupled thinking.
+
+## Chewing the CRUD
+
+A survey across the modern API space will reveal a coupled pattern for design that is the norm, albeit a limited one. The first design step in this pattern is usually to carefully craft a URL that can convey to the user the meaning of the content. The next step is to pick the HTTP method the users will use in interacting with that URL. The thought, design, and creativity is coupled to the HTTP implementation in this process, and pushes the question "What does my API do?" to subsequent steps in the design process.
+
+This approach maps directly into the concept of CRUD, which is frequently used to describe this kind of interface design. The term CRUD is from the area of database and encompasses the actions you take on persistent data. It stands for "Create, retrieve, update, and delete," which are common operations associated with interacting with persisted data. In the API world, these general semantics are matched up with HTTP methods, such as POST/PUT, GET, PUT/PATCH, and DELETE respectively. Using CRUD with HTTP APIs, a user performs these actions on a resource's URL.
```

Comment:
is the r actually retrieve. Whether right or wrong most use "read". You might want to comment this approach tunnels your data model structure into your URL effectively.